### DISCUSSION TITLE
Ech row marked as updated during first sync

### DISCUSSION THREAD

**Discussion by karolgt1** [Mimetis/Dotmim.Sync#894] [Q&A]
Hi, i have client db (on different server than server db) which was created by copying mdf file. When i'm calling GetEstimatedChangesCountAsync it shows number of changes (in server db) equal to the number of all rows from all tables taking part in sync process. From Ms SQL SMS both dbs looks the same. Due to big amount of data i want to avoid long time copying of data which is not needed. Maybe should i copy the data in some other way?

**Comment by Mimetis** [comment]
Hi @karolgt1,
You have mainly 2 ways to achieve this:

## DMS Snapshots

**DMS** can use "_snapshots_" to allow clients to bootstrap their sync with a lot of data coming from the server.
The idea is to create on the server a "_kind of_" directory containing all rows for all tables, that will be sent to the client as a "_zip_" file.

You can read more about this feature here: https://dotmimsync.readthedocs.io/Snapshot.html

## Sql Server to Sql Server : Using a backup

If clients and server are both **Sql Server**, you can use backups, but it's not included "out of the box" in **DMS**.
You need to do some extra job to be sure everything is correctly done and no rows are lost in the process.

In your case, as there is no row marked as tracked in the server database, as well as in the client database, **DMS** "_thinks_" it's a first synchronization.

First of all, if you are making a *_copy_* of the server database restored to the client, you need to be sure no rows will be added or removed to the server, before the client has restored the backup and synchronization has been initialized.
`DMS` is tracking changes on rows thanks to the tracking tables and the timestamp registered to each row.

So, if you want to use this kind of "_technic_" to start a new sync mechanism, based on a server backup, restored on a client, you can do it that way:

- Make a backup of the server database (if **DMS** is already configured, be sure you are not backup any tracking tables)
- Get the server database timestamp, that will be used on the client to get new / modified / deleted rows since the backup.
- Provision the server database with **DMS**, to be sure all new / modified or deleted rows are now tracked.
- Restore the backup on the client
- Provision the client with **DMS** to be sure all new / modified or deleted rows are now tracked
- Save the timestamp from the server, that will be used by client on next sync 
- Make a sync to get all rows from the backup to now.


Here is a piece of code, that you can use as exploration:

``` cs

setup = new SyncSetup("ProductCategory");
var localOrchestrator = new LocalOrchestrator(clientProvider);
var remoteOrchestrator = new RemoteOrchestrator(serverProvider);

// 1) Make a backup of the server database
// ....
// 

// 2) Provision server database
var serverScope = await remoteOrchestrator.ProvisionAsync(setup);

// 3) Get the timestamp to use on the client
var serverTimeStamp = await remoteOrchestrator.GetLocalTimestampAsync();

// Once you have done the first step (1), you need to make Provision (2) and get timestamp (3)
// really quickly to be SURE you don't have any new / modified or deleted rows in the server 
// database that are not in the backup and not tracked.
// Once (2) and (3) are done, you are safe and you have all the time for next steps...


// 4) Restore backup on client
// ...
//

// 5) Provision client
await localOrchestrator.ProvisionAsync(serverScope);

// 6) Get the local timestamp
var clientTimestamp = await localOrchestrator.GetLocalTimestampAsync();

// 7) Get scopeinfoclient
// ScopeInfoClient table contains all information fro the "next" sync to do (timestamp, parameters and so on ...)
var scopeInfoClient = await localOrchestrator.GetScopeInfoClientAsync();

// As we have some existing lines, we say it's not a new sync
scopeInfoClient.IsNewScope = false;

// Affecting the correct timestamp, the local one and the server one
scopeInfoClient.LastServerSyncTimestamp = serverTimeStamp;
scopeInfoClient.LastSyncTimestamp = clientTimestamp;
await localOrchestrator.SaveScopeInfoClientAsync(scopeInfoClient);

// We're done
// We can sync safely now
var agent = new SyncAgent(clientProvider, serverProvider, options);
var r = await agent.SynchronizeAsync(setup);

Console.Write(r.ToString());    
```
Let me know if it's what you are looking for

**Reply by karolgt1** [reply]
i forgot to mention that i'm using change tracker in my db. Is that fact changing any on these ways?

**Reply by Mimetis** [reply]
Yes.

But you know.... You can .... try by yourself ?

**Reply by karolgt1** [reply]
While going throw the second way (standard backup) i done Deprovisioning and then when i'm trying to Provision (just after backup was done) i'm getting this error:
"SqlException: Cannot find the type 'xxx_BulkType', because it does not exist or you do not have permission. "
Any idea?

**Reply by Mimetis** [reply]
hum, strange, can you share your code?
`xxx_BulkType` is the TVP Type generated for a table, that is used as input for the stored procedure.
This message should not appear during the `ProvisioAsync` call. Is it throw on the call to `ProvisionAsync` ?

**Reply by karolgt1** [reply]
yes, it throws while calling ProvisionAsync.
Apart of creation of setup (tables, columns and sync direction declaration) i'm calling this lines:
```
            var remote = new RemoteOrchestrator(sourceDb);
            var p = SyncProvision.ScopeInfo | SyncProvision.ScopeInfoClient |
                    SyncProvision.StoredProcedures | SyncProvision.TrackingTable |
                    SyncProvision.Triggers;
            await remote.DeprovisionAsync(p);
            var serverScope = await remote.ProvisionAsync(setup); 
```

**Reply by karolgt1** [reply]
i would like to mention that before doing Deprovision i was trying to call DropAll (to be sure that everything related with DMS was removed from db) but it was processing long time and after few minutes stopped execution.

**Reply by karolgt1** [reply]
btw. 'xxx_BulkType' => 'xxx' mens the name of the first table on tables list in syncSetup

**Reply by Mimetis** [reply]
You should be able to call `DropAllAsync()` on your `remoteOrchestrator`.
Basically, what you are doing is pretty the same as what is done by `DropAllAsync`.

I don't understand why `DropAllAsync()` is taking so much times, as it should be fast. May be a lock somewhere due to other connections opened to the server database?

That's being said, you can do the Backup of your server database, then restore to any client database, and **then on the new restored Client**, you can call the `DropAllAsync()` before making a new `ProvisionAsync()`

It makes more sense as I suppose you will have new clients over the time, and of course you don't want to deprovision the server.

**Reply by karolgt1** [reply]
ok, but think that i still need to provision a server?

**Reply by karolgt1** [reply]
while debugging i found that error occurs while creating 'xxx_bulkdelete' procedure, which inside yourself use xxx_BulkType which not exist

**Reply by Mimetis** [reply]
yes of course the server needs to be provisioned, otherwise no rows will be tracked
I don't know why some stored procedures don't have their TVP bulk type.
You should have play around the sql and DMS is lost

Try from scratch, and see if you are able to make a full scenario from scratch

**Reply by karolgt1** [reply]
when DMS is creating TVP bulk type? if they not exist, we should focus on creation of that

**Reply by Mimetis** [reply]
They are created when **DMS** is creating the related stored procedures

**Reply by karolgt1** [reply]
yes, but is it while calling ProvisionAsync or some other method? Im just wondering if is it ok that DMS doesn't create these TVP bul type by itself, maybe it is a bug?

**Reply by Mimetis** [reply]
yes, of course

**Reply by Mimetis** [reply]
Try to call `ProvisionAsync()` with the `overwrite` parameter
``` cs
var serverScope = await remoteOrchestrator.ProvisionAsync(setup, overwrite:true);
await localOrchestrator.ProvisionAsync(serverScope, ovewrite:true);

```

**Reply by karolgt1** [reply]
tried with this option and still the same error

**Reply by Mimetis** [reply]
I don't know
You need to share a sample repro, that I can use to debug your situation.

**Reply by karolgt1** [reply]
Hi, i found that the problem is in lack of permissions to call TVP bulks in procedure (i mean that sql user used by DMS doesn't have these permissions). Tried to add these permissions manually but they was again removed while Provisioning:
`            var serverScope = await remote.ProvisionAsync("DefaultScope", setup, overwrite: true);`
In some cases when i was not using 'overwrite: true' the provisioning process finshed correctly.
What is the problem? Why it clean these permissions?

**Reply by karolgt1** [reply]
i will simplify it to a question, what permissions for sql user (used by DMS) are needed on TVP bulks to use Provision method with overwrite as true ?

**Reply by Mimetis** [reply]
Are you owner of the database ?

`overwrite: true` will basically delete then create again the **DMS** stuff.

I guess your user has permissions to create things but not delete them ?

**Reply by Mimetis** [reply]
You need to be able to CREATE / DELETE / ALTER tables, stored procedures, triggers, table value paramter
You need to be able to EXEC stored procedures
You need to be able to DISABLE constraints in some circumstances

Honestly, you should use a user that have enough rights to manage the database

**Reply by karolgt1** [reply]
maybe do you have list of all permissions which sql user need to have? If no i think that it would be valuable to have this information in documentation

**Reply by Mimetis** [reply]
**DMS** needs to be able to **create / delete / alter tables, stored procedures, triggers** and so on...
Depending on the options you are using, **DMS** could potentially needs to disable constraints...

I don't know exactly the list of permissions needed, but I think that using a user with the **db_owner** role is required.
If you need a detailed list of permissions, it should require some investigations.
I will probably look into it, but not sure in the near future.

**Reply by karolgt1** [reply]
Next question, is it possible to configure the schema name, within which all the objects (TVPs, procedures, tables etc.) will be created?

**Reply by karolgt1** [reply]
I mean new different schema than the one which contains tables with data to sync

**Reply by Mimetis** [reply]
nop
